packages/remix-react/fog-of-war.ts

```diff
@@ -0,0 +1,259 @@
+    nextPaths: new Set<string>(),
+    knownGoodPaths: new Set<string>(),
+    known404Paths: new Set<string>(),
+  };
+
+  return {
+    enabled: true,
+    patchRoutesOnMiss: async ({ path, patch }) => {
+      if (fogOfWar!.known404Paths.has(path)) return;
+      await fetchAndApplyManifestPatches(
+        [path],
+        fogOfWar!.known404Paths,
+        manifest,
+        routeModules,
+        future,
+        isSpaMode,
+        basename,
+        patch
+      );
+    },
+  };
+}
+
+export function useFogOFWarDiscovery(
+  router: Router,
+  manifest: AssetsManifest,
+  routeModules: RouteModules,
+  future: FutureConfig,
+  isSpaMode: boolean
+) {
+  React.useEffect(() => {
+    // Don't prefetch if not enabled or if the user has `saveData` enabled
+    if (
+      !isFogOfWarEnabled(future, isSpaMode) ||
+      navigator.connection?.saveData === true
+    ) {
+      return;
+    }
+
+    // Register a link href for patching
+    function registerPath(path: string | null) {
+      let { knownGoodPaths, known404Paths, nextPaths } = fogOfWar!;
+      if (path && !knownGoodPaths.has(path) && !known404Paths.has(path)) {
+        nextPaths.add(path);
+      }
+    }
+
+    // Fetch patches for all currently rendered links
+    async function fetchPatches() {
+      fogOfWar?.controller?.abort();
+
+      let lazyPaths = getFogOfWarPaths(fogOfWar!, router);
+      if (lazyPaths.length === 0) {
+        return;
+      }
+
+      try {
+        fogOfWar!.controller = new AbortController();
+        await fetchAndApplyManifestPatches(
+          lazyPaths,
+          fogOfWar!.known404Paths,
+          manifest,
+          routeModules,
+          future,
+          isSpaMode,
+          router.basename,
+          router.patchRoutes,
+          fogOfWar!.controller.signal
+        );
+      } catch (e) {
+        console.error("Failed to fetch manifest patches", e);
+      }
+    }
+
+    // Register and fetch patches for all initially-rendered links
+    document.body
+      .querySelectorAll("a[data-discover]")
+      .forEach((a) => registerPath(a.getAttribute("href")));
+
+    fetchPatches();
+
+    // Setup a MutationObserver to fetch all subsequently rendered links
+    let debouncedFetchPatches = debounce(fetchPatches, 100);
+
+    function isElement(node: Node): node is Element {
+      return node.nodeType === Node.ELEMENT_NODE;
+    }
+
+    let observer = new MutationObserver((records) => {
+      records.forEach((r) => {
+        [r.target, ...r.addedNodes].forEach((node) => {
+          if (!isElement(node)) return;
+          if (node.tagName === "A" && node.getAttribute("data-discover")) {
+            registerPath(node.getAttribute("href"));
+          } else if (node.tagName !== "A") {
+            node
+              .querySelectorAll("a[data-discover]")
+              .forEach((el) => registerPath(el.getAttribute("href")));
+          }
+          debouncedFetchPatches();
+        });
+      });
+    });
```

Comment:
This logic detects newly rendered links on navigations and internal route state changes (i.e., click this button to show a link) and batches them up via the debounced fetch